core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -807,7 +807,11 @@ public InodeDirectory getRoot() {
           currentInodeDirectory.getId(), pathComponents[k], missingDirContext);
       if (currentInodeDirectory.isPinned() && !newDir.isPinned()) {
         newDir.setPinned(true);
-        newDir.setMediumTypes(new HashSet<>(currentInodeDirectory.getMediumTypes()));
+        if (!currentInodeDirectory.getMediumTypes().isEmpty()) {
+          newDir.setMediumTypes(new HashSet<>(currentInodeDirectory.getMediumTypes()));
+        } else {
+          newDir.setMediumTypes(Collections.emptySet());
+        }
```

Comment:
This is a lot of boilerplate code at the caller site. Instead of doing this, we can modify the `setMediumTypes` like below:

Instead of just setting the field to whatever is given, we do the copy internally.
```
setMediumTypes(Set<String> set) {
  if (set.isEmpty()) {
    mMediumTypes = Collections.emptySet();
  } else {
    mMediumTypes = new HashSet(set);
  }
}
```

So the caller will do
```
// Before
newDir.setMediumTypes(new HashSet<>(currentInodeDirectory.getMediumTypes()));
// After
newDir.setMediumTypes(currentInodeDirectory.getMediumTypes());
```

Above should be a better way. I'm just slightly concerned that it becomes different from a pure "setter", and therefore should have a better name to reflect the underlying copy, like `copyMediumTypes()`.

Reply:
How about we declare and reorganize mediumTypes as immutable?
So that we can handle this logic in two parts;
1- `getMediumTypes()` returns a reference to existing immutable reference (Collections.emptySet() for empty list)
2- `setMediumTypes()` or any add/remove/update method always create a new map.

Otherwise, I'm still in favor of handling this in `getMediumTypes()`, (Or `copyMediumTypes()`?), and make the setter as plain as possible.

Reply:
I'm in favor of this but i'm not sure I fully understood:
```
So that we can handle this logic in two parts;
1- getMediumTypes() returns a reference to existing immutable reference (Collections.emptySet() for empty list)
2- setMediumTypes() or any add/remove/update method always create a new map.
```

```
getMediumTypes() {
  // Just return the ref because it's immutable
  // The current interface doesn't say anything about immutability and because of the setter, we can't make this "final" :(
  // As of now, all callers are read-only
  return mMediumTypes;
}
setMediumTypes(Set<> types) {
  // Handle copy internally to ensure immutability
  if (types.isEmpty()) {
    mMediumTypes = Collections.emptySet();
  } else {
    mMediumTypes = new HashSet<>(types);
  }
}
// There are no add/remove/update methods in the interface so no need to worry
// The setter is the only update entrace
// Update paths include setAttribute and reading from journal etc
```

Reply:
One step further, the mutability of this `Set<String>` does not mean that much and we are not using it at all. If we know all possible medium types already these `Set`s can well be singletons and `EnumSet`s.

Currently we don't have a good definition of what the medium types are. The user can define anything not limited by "MEM,SSD,HDD". There's nothing we explicitly do to check the medium types match between master and all workers. Taking more steps back, setting this HashSet is currently only used by feature "pinning to specified medium". The overhead is really non-trivial for some feature that is rarely used.

Reply:
Moving to a separate thread for discussion https://github.com/Alluxio/alluxio/issues/14820

Reply:
I think EnumSet is a good choice here. We just need to fix the medium type candidate to be one of MEM, SSD, and HDD. I have never seen anything else used anyways.   If we want flexiblility, BitSet could be used. I agree that HashSet is too expensive as a per-inode construct right now.

Reply:
Jiacheng, right. That's what I meant.

For enum based representation of medium types, I think you might wanna see this section as well. I also think that's much better.
https://docs.google.com/document/d/1uAPy3aspGwzJmNZ_oUiFxLIdjGJD5-s5k0aRNUC4iqA/edit#heading=h.296qlbvti652

Reply:
I tried to convert MEDIUM to an enum and everything else will start being EnumSet. This is the implementation: https://github.com/Alluxio/alluxio/pull/15223

Let's compare this PR with #15223 and choose one. @yuzhu @ggezer

Reply:
On 2nd thought, moving to `ImmutableSet` in both member variable and the method signature answers our question on "Who does the defensive copy?" @ggezer 

Now the answer is, the caller does the copy and everything is now immutable explicitly and the signature makes the usage fool-proof enough. Using `ImmutableSet` gives us the type safety, in contrast with `unmodifiableSet` which is only `Set`. 

This pattern can be applied to many other internal collection-like member fields, if any.